### PR TITLE
FOUNDATIONS: Retrieve Comment By Id

### LINKED ISSUE #67: FOUNDATIONS: Retrieve Comment By Id

### PR DESCRIPTION
**Pull request by cjdutoit**
closes #67

### CODE REVIEW

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                    expectedCommentDependencyException))),
+                        Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowServiceExceptionOnRetrieveByIdIfDatabaseUpdateErrorOccursAndLogItAsync()
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs:60]: If**Service**Error

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.Comments
+{
+    public partial class CommentServiceTests
+    {
+        [Fact]
+        public async Task ShouldRetrieveCommentByIdAsync()
+        {
+            // given
+            Guid randomCommentId = Guid.NewGuid();
+            Guid inputCommentId = randomCommentId;
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs:23]: inputCommentId should be from the randomComment.

```csharp
Guid inputCommentId = randomComment.Id

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+            // then
+            actualComment.Should().BeEquivalentTo(expectedComment);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectCommentByIdAsync(inputCommentId),
+                    Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs:44]: As per the dependency hierarchy, datetime is before logging in the happy path.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            var invalidCommentId = Guid.Empty;
+
+            var invalidCommentException =
+                new InvalidCommentException();
+
+            invalidCommentException.AddData(
+                key: nameof(Comment.Id),
+                values: "Id is required");
+
+            var expectedCommentValidationException = new
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs:30]: Move the new keyword to next line.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                broker.SelectCommentByIdAsync(It.IsAny<Guid>()),
+                    Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowNotFoundExceptionOnRetrieveByIdIfCommentIsNotFoundAndLogItAsync()
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs:56]: This test should be under Validations.RetrieveById.cs file.
@hassanhabib Please correct if I am wrong.

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs (hunk 1 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs]: @ShriHumrudha have renamed test method

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs (hunk 2 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs]: @ShriHumrudha have updated this

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs (hunk 3 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs]: @ShriHumrudha have amended the order

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs (hunk 5 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs:56]: image
@ShriHumrudha the filename you suggest looks the same as its location?

Comment (ShriHumrudha) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs:56]: My bad. Thank you for confirming.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                broker.SelectCommentByIdAsync(It.IsAny<Guid>()),
+                    Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogCritical(It.Is(SameExceptionAs(
+                    expectedCommentDependencyException))),
+                        Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs:55]: Please change the hierarchy here - Storage, Logging, DateTime. Also, verify in other tests too.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+{
+    public partial class CommentServiceTests
+    {
+        [Fact]
+        public async Task ShouldRetrieveCommentByIdAsync()
+        {
+            // given
+            Guid randomCommentId = Guid.NewGuid();
+            Comment randomComment = CreateRandomComment();
+            randomComment.Id = randomCommentId;
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs:24]: I am still wondering if we need this. We are generating a random comment, then can we use the random comment's id as the inputCommentId?

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs (hunk 7 above)
Comment (hassanhabib) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs:24]: @ShriHumrudha I agree - this is not needed.
@cjdutoit you don't need to reassign the id. all we care about is that when a particular random id is passed an object comes back from the broker. that's all.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs (hunk 6 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.RetrieveById.cs:55]: @ShriHumrudha, so just to confirm, you only want the order to be storage, datetime, logger on the happy path and then storage, logger, datetime on the exception tests as I changed the order around based on the previous request?

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs (hunk 3 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs]: @ShriHumrudha, I have updated this

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs (hunk 4 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Validations.RetrieveById.cs]: @ShriHumrudha, I have updated this

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs (hunk 7 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Logic.RetrieveById.cs]: @ShriHumrudha, I have updated this

**Pull request merged**